### PR TITLE
Add pin feature for messages

### LINKED ISSUE #1905: Pinned Messages
https://github.com/status-im/status/issues/11

User Story
As a user, I want to pin or star messages in public channels for users of those channels to access them easily and quickly.

https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?node-id=4119%3A2

### PR DESCRIPTION
**Pull request by jrainville**
Fixes #1905 

Put as draft because it is waiting for the status-go PR (https://github.com/status-im/status-go/pull/2180)  to be merged and also because it is missing the status-go signal. Also, startWallet was removed from status-go, so I commented it for now, but I'll change it once Richard has a PR.

Adds the Pin and Unpin feature to non-public channels.

Adds a popup to see the pinned messages.

They also work for communities

Missing stuff:
- The API doesn't give the person who pinned the message, so I couldn't follow the design on this
- There is currently no signal when a pin or unpin is made, so it doesn't update automatically.
- Changes to startWallet because of the breaking change in status-go

image

image

### CODE REVIEW

File: src/status/libstatus/wallet.nim [hunk 1]
```diff
@@ -84,7 +84,8 @@ proc validateMnemonic*(mnemonic: string): string =
   result = $status_go.validateMnemonic(mnemonic)
 
 proc startWallet*(watchNewBlocks: bool) =
-  discard status_go.startWallet(watchNewBlocks) # TODO: true  to watch trx
+  discard
+  # discard status_go.startWallet(watchNewBlocks) # TODO: true  to watch trx
```
Comment (jrainville) [inline src/status/libstatus/wallet.nim:88]: This is no longer in status-go. It was a breaking change in the latest status-go. Richard is working on a fix, but it was easier for me for now to just comment it.

File: src/app/chat/views/message_list.nim [hunk 2]
```diff
@@ -231,9 +243,11 @@ QtObject:
   proc add*(self: ChatMessageList, message: Message) =
     if self.messageIndex.hasKey(message.id): return # duplicated msg
 
+    debug "New message", chatId = self.id, id = message.id, text = message.text
```
Comment (jrainville) [inline src/app/chat/views/message_list.nim:246]: Commenting here to rembeber to remove this

**Comment by anastasiyaig** [comment]
1. 3 buttons option is shown for **public** chats with no actions, when clicking it -> i cant close the menu, it will be shown forever

Possible expected result: remove 3 dots from public channels



2. Pin icon is too small 



3. Pinned **by** is missing (known limitation )

4. Message(s) does not reflect the count of messages, so it is always Message and not a plural. Also the representation of list of pinned messages is in yellow and quite messy


5. no reference to Pinned messages in Group chat profile



6. context menu is glued to a particular part of the screen, so resizing the window breaks it from the 3 dots


7. wrong hover on Pin / Unpin

**Comment by simonricoo** [comment]
7. Is just not updated design from the past, so the hover is ok :) It came with the latest redesign. However the length of the dropdown is quite long and **Pin** and **Copy** icons are blurred :((((((

**Comment by jrainville** [comment]
> Possible expected result: remove 3 dots from public channels, add possibility to close context menu by clicking out from the menu area

I know this was not part of the design, but I find that it makes sense to be able to see the other options using the three dots. So being able to copy the text from there as well as the right-click. Anyway, I do hide the "Pin" button when it's a public chat

**Comment by jrainville** [comment]
@anastasiyaig I fixed the issues you reported. Thanks for the thorough testing.

Some of them I couldn't fix or I didn't for various reasons, here they are:
- I added the `Pinned by` part, BUT, there is an issue in Status-go where if you close the app and reopen, the pinned by now shows the author of the message. I opened the issue here: https://github.com/status-im/status-go/issues/2234
- the point 6, where the menu is glued to the part of the screen should be opened in another issue, since it's something that affect all popup menus and not just the pin one.
- like Simon said, 7 is just the design not being up to date

@simonricoo sadly, I can't fix the rounded corner problem. The workaround we usually use to have different corners doesn't work if the color is translucid, because we can see the other rectangle underneath the first one.

**Comment by anastasiyaig** [comment]
@jrainville awesome work , thanks for putting an effort to this!

i logged the menu issue separately https://github.com/status-im/status-desktop/issues/2584, also thanks for opening the issue on go side! I checked -> it happens for communities channels and group chats, but for 1 x1 is okay 

There are small points which we probably can fix easily:

- increase click area for this menu (3 dots) -> for now there are places which highlight the button (meaning it is active) but clicking does not make any effect



- menu (3 dots) overlaps with pinned sticker message

Also, due to https://github.com/status-im/status/issues/11#issuecomment-773982069 unpin should be done:

1. by any admin of group chat and in this PR anyone can unpin no matter if he is admin or not
2. by admin in community channels and in this PR anyone can unpin no matter if he is admin or not

Can we please fix that as well?

**Comment by jrainville** [comment]
@anastasiyaig thanks again for the testing.

I fixed the overlap issue and also made it so that only admins can pin in communities and group chats

**Comment by anastasiyaig** [comment]
@jrainville 

1. The clicking area is still small with these 3 dots , so i still need to find exact place where to click it to get the options (please pay attention to the Pinned messages pop-up)

Screenshot 2021-05-24 at 12 45 55


2. Pinned message is still overlapping the other message

3. Pinned messages are not shown for non-admin **in a group chat**, it is kind of regression since it was there in previous build

Screenshot 2021-05-24 at 12 24 36

4. Group chat -> profile -> pinned messages throws an error
`qrc:/app/AppLayouts/Chat/components/GroupInfoPopup.qml:214: ReferenceError: pinnedMessagesPopupComponent is not defined`
5. Group chat -> profile -> pinned messages misses the icon
Screenshot 2021-05-24 at 12 31 34
6. Communities -> manage community: no pinned messages view option at all https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?node-id=5971%3A42716

image

7. also with point 6 we need zero screen ( 0 pinned messages)
image

**Review by richard-ramos — APPROVED** [review]
very nice!

the only change it requires is to update the status-go version it is using.

**Comment by jrainville** [comment]
@anastasiyaig I fixed the issues again.

Details about some of them:

1. I couldn't reproduce this one on Linux nor on Windows. Could you show a video? Does it only happen in the first element of the list or all of them?

2. this one is very hard to fix. I fixed it last time, but apparently my fix only works when you reopen the app. If you send a new message, then the fix doesn't work anymore (it doesn't calculate the Z index of the new element correctly for some reason). Since I couldn't fix it easily today, I opened a new issue to fix it later since I don't think it should be a blocker to merge the PR https://github.com/status-im/status-desktop/issues/2598

3. This is not an issue with this PR, it is either an issue in status-go or desktop as a whole, we'd need to check. Basically, as you can see on the right screen of your screenshot, the pinned messages are **before** the lest account joined the group chat. So the left account doesn't see (or doesn't have access?) to the messages. That's why it doesn't have the pinned messages buttons. If you add a new message and pin a new message after the left account joined, you'd see the button then. It's probably an issue of desktop where we don't fetch past messages when joining a group

6. The button is in the community channel edit popup, which is not merged yet. I added an issue to add the button later: https://github.com/status-im/status-desktop/issues/2597

**Comment by jrainville** [comment]
Rebased and updated status-go to match master

**Comment by anastasiyaig** [comment]
lets merge the PR and address the issues with backend separately

**Pull request merged**